### PR TITLE
config, api: Add Service Config

### PR DESCRIPTION
**Pull request by CabinfeverB**
Signed-off-by: Cabinfever_B <cabinfeveroier@gmail.com>

### What problem does this PR solve?

Issue Number: Ref #4480

### What is changed and how does it work?
Add an independent config for audit and rate limit in the future

```commit-message
add service config
```

### Check List

Tests

- Unit test
- Integration test

Code changes

- Has configuration change
- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/api/config.go [hunk 1]
```diff
@@ -151,6 +151,8 @@ func (h *confHandler) updateConfig(cfg *config.Config, key string, value interfa
 			return errors.Errorf("cannot update config prefix %s", kp[0])
 		}
 		return h.updateReplicationModeConfig(cfg, kp[1:], value)
+	case "service":
```
Comment (rleungx) [inline server/api/config.go:154]: Do we have a better naming?

File: server/api/router.go [hunk 2]
```diff
@@ -282,6 +282,10 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 	registerFunc(apiRouter, "/admin/persist-file/{file_name}", adminHandler.SavePersistFile, setMethods("POST"), setAuditBackend(localLog))
 	registerFunc(clusterRouter, "/admin/replication_mode/wait-async", adminHandler.UpdateWaitAsyncTime, setMethods("POST"), setAuditBackend(localLog))
 
+	selfProtectionHandler := newSelfProtectionHandler(svr, rd)
+	registerFunc(apiRouter, "/self_protection/config", selfProtectionHandler.GetSelfProtectionConfig, setMethods("GET"))
```
Comment (AndreMouche) [inline server/api/router.go:286]:
```suggestion
	registerFunc(apiRouter, "/self-protection/config", selfProtectionHandler.GetSelfProtectionConfig, setMethods("GET"))
```

File: server/api/router.go [hunk 3]
```diff
@@ -282,6 +282,10 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
...
+	registerFunc(apiRouter, "/self_protection/config", selfProtectionHandler.SetSelfProtectionConfig, setMethods("POST"), setAuditBackend(localLog))
```
Comment (AndreMouche) [inline server/api/router.go:287]:
```suggestion
	registerFunc(apiRouter, "/self-protection/config", selfProtectionHandler.SetSelfProtectionConfig, setMethods("POST"), setAuditBackend(localLog))
```

**Review by AndreMouche — COMMENTED** [review]
Could you please use hyphens (-) to improve the readability of URIs ? ref https://restfulapi.net/resource-naming/

File: server/api/router.go (hunk 2 above)
Comment (nolouch) [inline server/api/router.go:286]: Can we change `self_protection` to `service-middlewares or middlewares`? IMO, self-protection is better to be the project name, not in the code.

Comment (CabinfeverB) [reply server/api/router.go]: update

**Review by nolouch — COMMENTED** [review]
Overall LGTM. cc @rleungx

**Review by rleungx — COMMENTED** [review]
The rest LGTM

File: server/api/service_middleware.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+func newServiceMiddlewareHandler(svr *server.Server, rd *render.Render) *serviceMiddlewareHandler {
+	return &serviceMiddlewareHandler{
+		svr: svr,
+		rd:  rd,
+	}
+}
+
+// @Tags service_middleware
```
[the remark this answers is not in the document]
Comment (CabinfeverB) [reply server/api/service_middleware.go:45]: > better to keep the style with other interfaces

How about changing all interfaces to Snake case？ I have changed `service_gc_safepoint` and `min_resolved_ts`

**Review by rleungx — APPROVED** [review]

File: server/api/service_middleware.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+		return
+	}
+
+	conf := make(map[string]interface{})
+	if err := json.Unmarshal(data, &conf); err != nil {
+		h.rd.JSON(w, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	for k, v := range conf {
```
Comment (AndreMouche) [inline server/api/service_middleware.go:78]: what if the conf is empty?

File: server/api/service_middleware_test.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+func (s *testServiceMiddlewareSuite) TearDownSuite(c *C) {
+	s.cleanup()
+}
+
+func (s *testServiceMiddlewareSuite) TestConfigAudit(c *C) {
+	addr := fmt.Sprintf("%s/service-middleware/config", s.urlPrefix)
+	ms := map[string]interface{}{
+		"enable-audit": "true",
+	}
```
Comment (AndreMouche) [inline server/api/service_middleware_test.go:55]: How about add a case with an empty body or  `ms := map[string]interface{}{}`

**Review by AndreMouche — COMMENTED** [review]
Rest LGTM

File: server/api/service_middleware_test.go (hunk 6 above)
Comment (CabinfeverB) [reply server/api/service_middleware_test.go:55]: update

File: server/api/service_middleware.go (hunk 5 above)
Comment (CabinfeverB) [reply server/api/service_middleware.go:82]: add a check

**Review by AndreMouche — APPROVED** [review]
LGTM

**Pull request merged**